huffman/src/table.hpp

```diff
@@ -67,27 +105,23 @@ class table
 
     std::ranges::sort(table_);
 
+    // precondition
     assert(
         std::ranges::unique(
-            table_, [](auto& x, auto& y) { return x.symbol == y.symbol; })
+            table_, {}, [](const auto& elem) { return elem.symbol; })
             .empty() and
-        "`frequencies` cannot contain duplicate symbols");
-
-    while (table_.front().node_size() != table_.size()) {
-      table_.front().join_with_next();
+        "a `table` cannot contain duplicate symbols");
 
-      const auto last = table_.data() + table_.size();
-      const auto has_higher_freq =
-          [f = table_.front().frequency()](const auto& n) {
-            return n.frequency() > f;
-          };
+    const auto frequencies = std::views::transform(
+        table_, [](const auto& elem) { return elem.frequency(); });
+    [[maybe_unused]] const auto total_freq =
+        std::accumulate(std::cbegin(frequencies), std::cend(frequencies), 0UZ);
 
-      auto lower = table_.front().next();
-      auto upper = find_node_if(lower, last, has_higher_freq);
+    encode_symbols();
 
-      // re-sort after creating a new internal node
-      std::rotate(&table_.front(), lower, upper);
-    }
+    // postcondition
+    assert(total_freq == table_.front().frequency());
+    std::ranges::reverse(table_);
```

Comment:
add a comment about why

Reply:
I'll add this later:

If front().frequency() does not equal total_freq, we have skipped over nodes during construction of the implicit Huffman tree.

Reply:
Or you feel free to :)

Reply:
I meant comment about why the reverse, not about the assertion.

Reply:
done